### DISCUSSION TITLE
Adding cover to FP

### DISCUSSION THREAD

**Discussion by AlekeiPL** [ExperienceLovelace/ha-floorplan#270] [Q&A] [answered]
Hi, first of all, excellent job !!
For two days I have been trying to add a picture of a closed roller blind, unfortunately, no matter what I do, I am not able to add. Tried to do overlay, area and nothing works. The only thing I figured out was that I can use the entitle when I click where the blind should be. I made a render when the shutter is closed Maybe some of you will be nice enough to show me where I'm making a mistake
Here is my config:
```
type: custom:floorplan-card
style: |
  ha-card {
      max-width: 150vh
      margin: 0 auto;
      background: none;
      box-shadow: none;
  }
config:
  image: /local/floorplan/dom52.svg
  stylesheet: /local/floorplan/floorplan.css
  rules:
    - entity: media_player.sony_bravia_tv_2
      tap_action: toggle
      state_action:
        action: call-service
        service: floorplan.class_set
        service_data: media-player-tv media-player-${entity.state}
    - entities:
        - light.lampa_4
        - light.lampa_1
        - light.lampa_3
        - light.salon_3
        - light.kuchnia_3
        - light.hue_color_lamp_1
        - light.lazienka
        - light.sypialnia
      tap_action: false
      state_action:
        action: call-service
        service: floorplan.class_set
        service_data: light-${entity.state}
    - entities:
        - light.lampa_4
        - light.lampa_1
        - light.lampa_3
        - light.salon_3
        - light.kuchnia_3
        - light.hue_color_lamp_1
        - light.lazienka
        - light.sypialnia
      tap_action:
        action: toggle
      double_tap_action:
        action: more-info
      state_action:
        action: call-service
        service: floorplan.style_set
        service_data:
          element: ${entity.entity_id.replace('light.', 'light_overlay.')}
          style: |
            >
            if( entity.state !== "on" )
                return "display: none;";
            let hue = 0;
            let sat = 0;
            if( entity.attributes.hs_color )
            {
                hue = entity.attributes.hs_color[0];
                sat = entity.attributes.hs_color[1];
            }
            if( sat < 10 )
            {
              return `
                display: block;
                filter:
                  brightness(calc( ${entity.attributes.brightness} / 255));`
            }
            return `
              display: block;
              filter:
                sepia(100%)
                hue-rotate(calc( ${hue}deg - 55deg ))
                saturate(calc( ${sat}% * 2 ))
                brightness(calc( ${entity.attributes.brightness} / 255));
            `
    - entities:
        - cover.roleta
      tap_action: false
      state_action:
        action: call-service
        service: floorplan.class_set
        service_data: cover-${entity.state}
    - entities:
        - cover.roleta
      tap_action:
        action: toggle
      double_tap_action:
        action: more-info
      state_action:
        action: call-service
      service: floorplan.style_set
      service_data:
        element: ${entity.entity_id.replace('cover.', 'cover_overlay.')}
```

![dom33](https://user-images.githubusercontent.com/92952299/203159754-308b7c9a-426d-44b1-8143-fea315eec62f.svg)

**Comment by OzGav** [comment]
We can’t read your code easily as you haven’t formatted it. Please edit your post and highlight each code block and press the <> button. 

You seem to have two rules for the entity `cover.roleta` which isn’t right

**Reply by exetico** [reply]
Consider to share the SVG, too 😊! 

And Thank you for the kind words. 

Use these things 😎:
`\``yaml

Cool stuff here

`\``

Like so:
![Screenshot_20221121_221219_Chrome](https://user-images.githubusercontent.com/3549445/203159225-8a29d03f-308a-4e35-af8b-df033e3941ae.jpg)

**Reply by AlekeiPL** [reply]
> We can’t read your code easily as you haven’t formatted it. Please edit your post and highlight each code block and press the <> button.
> 
> You seem to have two rules for the entity `cover.roleta` which isn’t right

i tried to do it in the same principle as light


Thx, for answering! I'm not sure which version is closer to working, so I left two

**Reply by OzGav** [reply]
Yeah it isn’t correct to have two entries like that. Typically you would use either `style_set` or `class_set` not both.

**Reply by OzGav** [reply]
And your `style_set` call for `cover.roleta` isn’t correct as it needs to define the CSS properties you want to assign to the `cover.roleta` element

**Reply by OzGav** [reply]
Look at the home example and `light.garage` for how to change an image https://experiencelovelace.github.io/ha-floorplan/docs/example-home/

**Reply by AlekeiPL** [reply]
```
@keyframes fade-in {
	from {
		opacity: 0;
	}
	to: {
		opacity: 1;
	}
}


svg, svg * {
  pointer-events: all !important;
  cursor: auto !important;
}

svg path {
  fill: inherit;
  stroke: inherit;
  stroke-width: inherit;
}


#areas {
  fill: none;
}

#light-overlays image {
  display: none;
  mix-blend-mode: lighten;
  animation: 0.75s fade-in linear forwards;
}

#light-overlays image.light-on {
  display: block;
}
}

#cover-overlays image {
  display: none;
  mix-blend-mode: lighten;
  animation: 0.75s fade-in linear forwards;
}

#cover-overlays image.cover-closed {
  display: block;
}


#areas .motion-on
{
 	fill: #ff00002e;
 	stroke: #ff000087;
 	stroke-width: 3px;
 	stroke-opacity: 1;
 	filter: unset;
 	animation: fade-in 1s infinite alternate;
}




#areas .light-on {
  stroke-width: 6px;
  stroke-opacity: 1;
  stroke: #a58ee1;
}

g path:active, g use:active, .light-on:active, g#entities > g:active path:not(.touch-area) {
	stroke: #9ed8ff99 !important;
	stroke-width: 4px;
	stroke-opacity: 0.7;
}

g#entites path, g#entites ellipse {
	stroke: var(--paper-item-icon-color);
	stroke-width: 1px;
	stroke-opacity: 1;
	fill: none;
}
g#entites .light-on {
	stroke: var(--paper-item-icon-active-color);
}

g#entites .media-player-tv.media-player-idle,
g#entites .media-player-tv.media-player-standby,
g#entites .media-player-tv.media-player-paused {
	stroke: var(--paper-item-icon-active-color);
	stroke-width: 5px;
}
```
I add cover in CSS just like lights.

Ok i will look.

**Reply by AlekeiPL** [reply]
Is this same SVG as main floorplan on its another one?
"image: /local/floorplan/examples/home/light_${entity.state}.svg"

**Reply by OzGav** [reply]
No the SVGs are small images of the light bulb

**Reply by AlekeiPL** [reply]
I added:

```
    - entity: cover.roleta
      element: cover.roleta
      state_action:
        action: call-service
        service: floorplan.image_set
        service_data:
          image: /local/floorplan/cover_${entity.state}.svg
          cache: true
        tap_action: toggle
       
```
Now i cant open cover but i can see "closed" picture 
![image](https://user-images.githubusercontent.com/92952299/203167539-4a149c66-35da-4fe0-991e-ef9e0a265dda.png)

**Reply by OzGav** [reply]
`tap_action` isn't indented correctly

**Reply by AlekeiPL** [reply]
It`s working!!!! Thank you very much guys!

**Reply by exetico** [reply]
Please share the final result, related to the original question 😊👍.

**Reply by AlekeiPL** [reply]
My problem was that I was figuring out too much how to do it.
All I had to do was use "service: floorplan.image_set" and set up the right name on img.svg block. Two days of torment, and thanks to you the problem turned out to be simple. Below the config, as if someone needs it. Thx again!
```
type: custom:floorplan-card
style: |
  ha-card {
      max-width: 150vh
      margin: 0 auto;
      background: none;
      box-shadow: none;
  }
config:
  image: /local/floorplan/dom54.svg
  stylesheet: /local/floorplan/floorplan.css
  rules:
    - entity: media_player.sony_bravia_tv_2
      tap_action: toggle
      state_action:
        action: call-service
        service: floorplan.class_set
        service_data: media-player-tv media-player-${entity.state}
    - entity: cover.roleta
      element: cover.roleta
      state_action:
        action: call-service
        service: floorplan.image_set
        service_data:
          image: /local/floorplan/cover_${entity.state}.svg
      tap_action:
        action: toggle
    - entities:
        - light.lampa_4
        - light.lampa_1
        - light.lampa_3
        - light.salon_3
        - light.kuchnia_3
        - light.hue_color_lamp_1
        - light.lazienka
        - light.sypialnia
      tap_action: false
      state_action:
        action: call-service
        service: floorplan.class_set
        service_data: light-${entity.state}
    - entities:
        - light.lampa_4
        - light.lampa_1
        - light.lampa_3
        - light.salon_3
        - light.kuchnia_3
        - light.hue_color_lamp_1
        - light.lazienka
        - light.sypialnia
      tap_action:
        action: toggle
      double_tap_action:
        action: more-info
      state_action:
        action: call-service
        service: floorplan.style_set
        service_data:
          element: ${entity.entity_id.replace('light.', 'light_overlay.')}
          style: |
            >
            if( entity.state !== "on" )
                return "display: none;";
            let hue = 0;
            let sat = 0;
            if( entity.attributes.hs_color )
            {
                hue = entity.attributes.hs_color[0];
                sat = entity.attributes.hs_color[1];
            }
            if( sat < 10 )
            {
              return `
                display: block;
                filter:
                  brightness(calc( ${entity.attributes.brightness} / 255));`
            }
            return `
              display: block;
              filter:
                sepia(100%)
                hue-rotate(calc( ${hue}deg - 55deg ))
                saturate(calc( ${sat}% * 2 ))
                brightness(calc( ${entity.attributes.brightness} / 255));
            `
```

**Comment by AlekeiPL** [comment]
Hi again! 
I don't want to make another post so I don't mess up, so I'm asking here. Can you upload animations to an svg file? 
```
    - entity: climate.klima
      element: climate.klima
      state_action:
        - service: floorplan.class_set
          service_data:
            class: '${(entity.state === "heat") ? "klimatyzacja" : ""}'
        - service: floorplan.image_set
          service_data:
            image: /local/floorplan/climate_${entity.state}.svg
      tap_action: more-info
```
I would like the image to wave, I tried different options in the CSS file and if I remove "image", the animation works, but with "image" it does not want to work. If someone would be so kind and give me the simplest example of animations, I'll be able to handle it further. Thank you for your answer 🙂

**Reply by pkozul** [reply]
Hi @AlekeiPL

Can you please share the floorplan SVG file, the climate SVG file(s), and the CSS you are using. We can then try and replicate the behaviour and see what the cause is.

Cheers,
Petar

**Reply by AlekeiPL** [reply]
![DOM](https://user-images.githubusercontent.com/92952299/204453092-979ba8a8-9eeb-42c9-ad4f-d952833783e1.svg)
SVG main file
![climate_heat](https://user-images.githubusercontent.com/92952299/204453164-bb34d011-e781-46a5-b632-1ef343c2cc4c.svg)
Climate SVG file
```
@keyframes fade-in {
	from {
		opacity: 0;
	}
	to: {
		opacity: 1;
	}
}


svg, svg * {
  pointer-events: all !important;
  cursor: auto !important;
}

svg path {
  fill: inherit;
  stroke: inherit;
  stroke-width: inherit;
}

.klimatyzacja {
  animation-name: klima;
  animation-duration: 5s;
  animation-iteration-count: infinite;
  animation-timing-function: linear;
  transform-origin: 50% 50%;
  transform-box: fill-box;
}

@keyframes klima {
	0% {
		animation-timing-function: ease-in;
		opacity: 1;
		transform: translateY(-45px);
	}

	24% {
		opacity: 1;
	}

	40% {
		animation-timing-function: ease-in;
		transform: translateY(-24px);
	}

	65% {
		animation-timing-function: ease-in;
		transform: translateY(-12px);
	}

	82% {
		animation-timing-function: ease-in;
		transform: translateY(-6px);
	}

	93% {
		animation-timing-function: ease-in;
		transform: translateY(-4px);
	}

	25%,
	55%,
	75%,
	87% {
		animation-timing-function: ease-out;
		transform: translateY(0px);
	}

	100% {
		animation-timing-function: ease-out;
		opacity: 1;
		transform: translateY(0px);
	}
}

#areas {
  fill: none;
}

#light-overlays image {
  display: none;
  mix-blend-mode: lighten;
  animation: 0.75s fade-in linear forwards;
}

#light-overlays image.light-on {
  display: block;
}


#areas .motion-on
{
 	fill: #ff00002e;
 	stroke: #ff000087;
 	stroke-width: 3px;
 	stroke-opacity: 1;
 	filter: unset;
 	animation: fade-in 1s infinite alternate;
}




#areas .light-on {
  stroke-width: 6px;
  stroke-opacity: 1;
  stroke: #a58ee1;

}

g path:active, g use:active, .light-on:active, g#entities > g:active path:not(.touch-area) {
	stroke: #9ed8ff99 !important;
	stroke-width: 4px;
	stroke-opacity: 0.7;
}

g#entites path, g#entites ellipse {
	stroke: var(--paper-item-icon-color);
	stroke-width: 1px;
	stroke-opacity: 1;
	fill: none;
}
g#entites .light-on {
	stroke: var(--paper-item-icon-active-color);
}

g#entites .media-player-tv.media-player-idle,
g#entites .media-player-tv.media-player-standby,
g#entites .media-player-tv.media-player-paused {
	stroke: var(--paper-item-icon-active-color);
	stroke-width: 5px;
}
```
CSS file

I used some ready-made script here to check if it works.

**Comment by pkozul** [comment, marked as answer]
Hi @AlekeiPL

I have managed to get it working. Rather than set the animation on the SVG itself, I have added a `<g>` element to simply act as a container for the SVG. The animation is then applied to this container.

![image](https://user-images.githubusercontent.com/2073827/204502414-00b572cc-7373-4a7a-af14-332a26ffb422.png)

Below is the YAML. As you can see, I added the line `element: climate.klima.container` to ensure the CSS class is set on the container `<g>` element, rather than the SVG itself.

```yaml
    - entity: climate.klima
      element: climate.klima
      state_action:
        - service: floorplan.class_set
          service_data:
            element: climate.klima.container
            class: '${(entity.state === "heat") ? "klimatyzacja" : ""}'
        - service: floorplan.image_set
          service_data:
            image: /local/floorplan/climate_${entity.state}.svg
      tap_action: more-info
```

Below is the updated SVG file which contains the `<g>` container element:

![light](https://user-images.githubusercontent.com/2073827/204501888-4e485762-0688-423e-8362-83d1b2583aef.svg)

The other way would be to remove the call to `class_set` in the YAML, and set the class in the SVG itself (assuming you never need to remove the CSS class). In that case, the `<g>` container element would be:

```svg
 <g class="klimatyzacja">
```

Can you try this and let us know how go?

**Reply by AlekeiPL** [reply]
Works like a dream! So if we want to animate something, it's enough to put it as a container folder in Inkscape?

**Reply by exetico** [reply]
The simple answer will be yes :)

**Reply by AlekeiPL** [reply]
Thanku for your time and help again! :)

**Reply by AlekeiPL** [reply]
Wait, I rushed a bit. Turning it on works, the animation works, but when I turn it off, the picture remains instead of disappearing.

**Reply by exetico** [reply]
You should have another image for the off-state - or a blank one, or else, hide that part of the SVG with opacity, and maybe even disable events (both with CSS)... If I understand you correctly ☺️.

The SVG file is requested for all stages (fx the on and off state).

But now that you've implemented a solution where both SVG files will be requested (on and off), I giesd you should go with another (empty) SVG file as the first try👍.

**Reply by AlekeiPL** [reply]
Ok I understand. SVG for every state!
This will be the last question and I will have the necessary knowledge to finish my FP. Is it possible to make a conditional image? I'd like to see when the door is open (not a problem now, thanks to you) but the image to show the status of the door would have to depend on whether the light is on or off. Is it doable?

**Reply by exetico** [reply]
Yes. At this point, you should check a few of the latest topics. The community is also here, for you to do your own research and searching, too ☺️👍. 

Here you can find some examples on how to combine two different entity states in ha-flooplan.

**Reply by exetico** [reply]
This should help you go in the right direction - at least a bit ;-)

https://github.com/ExperienceLovelace/ha-floorplan/discussions/264#discussioncomment-4171701

It's more advance, and you're going into a new type of logics.

Let us know, if you face any problems. But please, start by giving it a try.

**Reply by AlekeiPL** [reply]
Roger that! Thank You

**Reply by pkozul** [reply]
If you want to hide the image when `climate.klima` is switched off, then you can do it this way. We are using the `<g>` container element to control the visibility.

```yaml

    - entity: climate.klima
      element: climate.klima
      state_action:
        - service: floorplan.class_set
          service_data:
            element: climate.klima.container
            class: '${(entity.state === "on") ? "klimatyzacja-on" : "klimatyzacja-off"}'
        - service: floorplan.image_set
          service_data:
            image: /local/floorplan/climate_${entity.state}.svg
      tap_action: more-info
```

As you can see, the difference is that we now have a CSS class for each state. We simply use `display: none` to hide the image (container), and `display: block` to show it.

```css
.klimatyzacja-off {
  display: none;
}

.klimatyzacja-on {
  display: block;
  animation-name: klima;
  animation-duration: 5s;
  animation-iteration-count: infinite;
  animation-timing-function: linear;
  transform-origin: 50% 50%;
  transform-box: fill-box;
}
```

**Reply by AlekeiPL** [reply]
Thank you for explaining it so nicely and simply. Great job, I envy your knowledge 🙂
